### PR TITLE
Refactor deploy.Broker to use the client.Producer

### PR DESCRIPTION
**Pull request by tpantelis**
This requires changing the underlying `Ensure` functions to take client interfaces instead of `rest.Config` and moving other components under _pkg/internal_.  See individual commits for details.

Depends on #1705

### CODE REVIEW

**Review by skitt — APPROVED** [review]

File: pkg/broker/file.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		// TODO return reporter.Error(err, "error creating Kubernetes client")
+		return errors.Wrap(err, "error creating Kubernetes client")
+	}
+
+	data, err := newDataFrom(kubeClient, brokerNamespace, ipsecFile)
+	if err != nil {
+		// TODO return reporter.Error(err, "error initializing broker info")
```
Comment (Jaanki) [inline pkg/broker/file.go:51]: Remove this TODO

File: pkg/broker/file.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		return err
+	}
+
+	data.BrokerURL = restConfig.Host + restConfig.APIPath
+
+	newFilename, err := backupIfExists(InfoFileName)
+	if err != nil {
+		// TODO return reporter.Error(err, "error backing up the broker file")
```
Comment (Jaanki) [inline pkg/broker/file.go]: Ditto

File: cmd/subctl/deploybroker.go [hunk 3]
```diff
@@ -39,7 +42,14 @@ var deployBroker = &cobra.Command{
 	Use:   "deploy-broker",
 	Short: "Deploys the broker",
 	Run: func(cmd *cobra.Command, args []string) {
-		err := deploy.Broker(&deployflags, restConfigProducer, cli.NewReporter())
+		status := cli.NewReporter()
+		err := deploy.Broker(&deployflags, restConfigProducer, status)
+		if err == nil {
```
Comment (Jaanki) [inline cmd/subctl/deploybroker.go]: reworking the logic to:

```
err := deploy.Broker(&deployflags, restConfigProducer, status)
exit.OnError("Error deploying Broker", err)

configs, _ := restConfigProducer.ForCluster()
			err = broker.WriteInfoToFile(configs, deployflags.BrokerNamespace, ipsecSubmFile,
				stringset.New(deployflags.BrokerSpec.Components...), deployflags.BrokerSpec.DefaultCustomDomains, status)
		}
```

File: pkg/deploy/broker.go [hunk 4]
```diff
@@ -51,50 +50,45 @@ var ValidComponents = []string{component.ServiceDiscovery, component.Connectivit
 // Ignoring th cyclic complexity of Broker function because it is being refactored in
 // https://github.com/submariner-io/submariner-operator/pull/1717.
 //gocyclo:ignore
```
Comment (Jaanki) [inline pkg/deploy/broker.go]: unignore this check

**Review by Jaanki — CHANGES_REQUESTED** [review]

**Review by Jaanki — COMMENTED** [review]
For the commit `Modify cli.Status to log directly if status not set` I don't think this was necessary because there are cases where the status.Start/End happens in the calling functions.

File: pkg/broker/file.go (hunk 1 above)
Comment (tpantelis) [reply pkg/broker/file.go:51]: yeah I thought I had done that

**Comment by tpantelis** [comment]
> For the commit `Modify cli.Status to log directly if status not set` I don't think this was necessary because there are cases where the status.Start/End happens in the calling functions.

There are places in `deploy.Broker` where we return an error w/o invoking the reporter and let the caller log it via `exit.OnError.` I think we should be consistent and always notify the reporter. Also in those cases and others, we don't really need a Start/End block. This provides the flexibility to log a series of operations via a Start/End block or just log a single operation.

File: pkg/deploy/broker.go (hunk 4 above)
Comment (tpantelis) [reply pkg/deploy/broker.go]: yeah I meant to do that.

File: cmd/subctl/deploybroker.go (hunk 3 above)
Comment (tpantelis) [reply cmd/subctl/deploybroker.go]: Restructured it.

**Review by tpantelis — COMMENTED** [review]
Addressed comments with https://github.com/submariner-io/submariner-operator/pull/1717/commits/54415d5770860cb66a5c2bfee57591786e5e282c.

File: internal/exit/exit.go [hunk 5]
```diff
@@ -24,23 +24,23 @@ import (
 	"github.com/submariner-io/submariner-operator/pkg/version"
 )
 
-// OnError will print your error nicely and exit in case of error.
-func OnError(message string, err error) {
+// OnError exits in case of error.
+func OnError(err error) {
```
Comment (Jaanki) [inline internal/exit/exit.go:28]: We do want to print what the err is

File: cmd/subctl/deploybroker.go [hunk 6]
```diff
@@ -43,20 +43,20 @@ var deployBroker = &cobra.Command{
...
-		exit.OnError("Error creating client producer", err)
+		exit.OnError(status.Error(err, "Error creating client producer"))
 
-		status := cli.NewReporter()
 		err = deploy.Broker(&deployflags, clientProducer, status)
-		if err == nil {
-			err = broker.WriteInfoToFile(config, deployflags.BrokerNamespace, ipsecSubmFile,
-				stringset.New(deployflags.BrokerSpec.Components...), deployflags.BrokerSpec.DefaultCustomDomains, status)
-		}
+		exit.OnError(err)
```
Comment (Jaanki) [inline cmd/subctl/deploybroker.go:55]: exit.OnError(status.Error(.....))?

**Review by Jaanki — CHANGES_REQUESTED** [review]

File: internal/exit/exit.go (hunk 5 above)
Comment (tpantelis) [reply internal/exit/exit.go:28]: With the current usages the error is already printed - here we provide a convenience to exit if non-nil. If the caller wants to print the error they can call `WithMessage`.

File: cmd/subctl/deploybroker.go (hunk 6 above)
Comment (tpantelis) [reply cmd/subctl/deploybroker.go:55]: `deploy.Broker` has already logged the error via the `status`. Here we just want to exit with a non-zero code if `err` is  non-nil.

Comment (Jaanki) [reply cmd/subctl/deploybroker.go:55]: On second thought, `status` should end in the `pkg` itself and not propagate to `cmd/subctl`. This way any user using `broker` pkg will have a complete status implementation.

**Review by Jaanki — CHANGES_REQUESTED** [review]

**Review by Jaanki — APPROVED** [review]

**Pull request merged**